pupepat/utils.py

```diff
@@ -13,32 +13,92 @@
 
+#
+# configuraton for everything: image filter, cutout, solver guesses, etc.
+# this dictionary holds the default values. if --config-file is specified,
+# that file is read and this dict is updated. It's written to the --output-dir
+# as yaml file
+#
+# guess for the size of a defocused donut is 0.387" 4 FOCDMD
+#
+config = {
+    'SEP': {
+        'extract_SN_threshold': 10.0,
+        'mask_threshold_scale_factor': 10.0,
+        'min_area': 2000,
+        'bg_box_size': 128,
+    },
+    'cutout': {
+        'cutout_radius': 150,
+        'cutout_padding': 0.2 # percent additional margin added to each edge
+    },
+    'source_filters': {'bad_column_max': 400.0,
+                       'bad_column_min': 200.0,
+                       'edge_proximity_limit': 150.0,
+                       'focus_scale_factor': 200.0,
+                       'ellipticity_limit': 1.15,
+    },
+}
+
+def update_mapping(dest, source):
+    """Update dest dictionary with key,value pairs from source dictionary.
+    """
+    for key, value in source.items():
+        if isinstance(value, collections.Mapping):
+            dest[key] = update_mapping(dest.get(key, {}), value)
+        else:
+            dest[key] = value
+    return dest
+
 
 def offsets(x1, y1, x2, y2):
     return ((x2 - x1) ** 2.0 + (y2 - y1) ** 2.0) ** 0.5
 
 
-def estimate_bias_level(hdu):
+def get_bias_corrected_data_in_electrons(hdu):
     """
-    Estimate the bias level of image
+    Estimate the bias level of image and substract it from the image.
+    If bias estimate is negative, issue warning and scale data.
+
+    Uses median_absolute_deviation (MAD) to compute standard deviation.
+    Note: does not substract background from data to compute noise, as was done
+    prior to Sept.2018 parameter tuning.
+
     :param hdu: fits hdu with the image in question
 
-    :return: float bias_level
+    :return: np.array data_e
     """
+    gain = float(hdu[0].header['GAIN'])            # e-/count
+    read_noise_e = float(hdu[0].header['RDNOISE']) # e-/pixel
+    data_e = gain * hdu[0].data                    # counts to electrons
+
     # 1.48 here goes from median absolute deviation to standard deviation
-    noise = 1.48 * median_absolute_deviation(hdu[0].data - sep.Background(hdu[0].data.astype(np.float)))
-    gain = float(hdu[0].header['GAIN'])
-    read_noise = float(hdu[0].header['RDNOISE'])
-    bias_level = gain * np.median(hdu[0].data) - gain * gain * noise * noise + read_noise * read_noise
-    bias_level /= gain
-    return bias_level
+    noise_e = 1.48 * median_absolute_deviation(data_e)
+
+    estimated_bias_level_in_electrons = np.median(data_e) - noise_e * noise_e + read_noise_e * read_noise_e
+
+    # If the bias is negative, there's something wrong.
+    # Scaling the data here is a work-around for that.
+    if estimated_bias_level_in_electrons < 0:
+        noise_e = 1.48 * median_absolute_deviation(data_e)
+        sqrt_median_e = np.sqrt(np.median(data_e))
```

Comment:
Just a thought: when we have negative bias and suspect the gain is wrong, should we remove the incorrect gain from the data before we proceed?  (at line 81 we say `data_e = gain * hdu[0].data`). (maybe it doesn't matter because we're just rescaling the data anyway....)